crates/dbsp/src/operator/dynamic/balance/accumulate_trace_balanced.rs

```diff
@@ -0,0 +1,1306 @@
+};
+use tracing::info;
+
+circuit_cache_key!(BalancedTraceId<C: Circuit, B: IndexedZSet>(StreamId => (Stream<C, Option<Spine<B>>>, Stream<C, TimedSpine<B, C>>)));
+
+impl<C, B> Stream<C, B>
+where
+    C: Circuit,
+    B: IndexedZSet,
+{
+    /// Shard and integrate `self` using dynamic sharding policy selected by the balancer.
+    ///
+    /// Combines exchange, accumulator, and integral into a single operator. In steady state,
+    /// applies the selected sharding policy to the input stream. When the sharding policy changes,
+    /// retracts the previous content of the integral and re-distributes it using the new policy.
+    ///
+    /// # Returns
+    ///
+    /// The output of the accumulator and the integral.
+    ///
+    /// # Circuit
+    ///
+    /// ```text
+    ///                                      │
+    ///                                      │self
+    ///                                      ▼
+    ///                         ┌─────────────────────────┐
+    ///              ┌─────────►│RebalancingExchangeSender│◄───┐
+    ///              │          └────────────┬────────────┘    │
+    ///              │                       .                 │delayed_acc
+    ///              │                       .                 |
+    ///              │                       ▼                 │
+    ///              │               ┌────────────────┐      ┌──┐
+    ///              │               │ExchangeReceiver│      │Z1│
+    ///              │               └────────────────┘      └──┘
+    ///              │                       │                 ▲
+    ///              │                       │                 │
+    ///              │                       ▼                 │
+    ///              │             ┌──────────────────────┐    │
+    ///              │             │RebalancingAccumulator│────┘
+    /// delayed_trace│             └──────────────────────┘
+    ///              │                       │
+    ///              │                       │────────────────────┐
+    ///              ├──────────────────┐    │                    │
+    ///              │                  ▼    ▼                    │
+    ///      ┌───────┴─────────┐    ┌─────────────────────┐       │
+    ///      │AccumulateZ1Trace│◄───┤AccumulateTraceAppend│       │
+    ///      └─────────────────┘    └────────┬────────────┘       │
+    ///                                      │                    │
+    ///                                      │trace               │accumulator_stream
+    ///                                      │                    │
+    ///                                      ▼                    ▼
+    /// ```
+    ///
+    /// The two feedback streams (delayed_acc and delayed_trace) are used by RebalancingExchangeSender to
+    /// retract previously accumulated content of the integral when the sharding policy changes. The accumulator
+    /// in particular is used to rebalance mid-transaction.
```

Comment:
I think this is very interesting: we actually use the incremental computation mechanism to update the internal state itself, taking advantage of the fact that the state is a Z-set. The rest of the circuit does not need to do anything special - the invariant is that the none of the Z-sets represented by some streams in the circuit do not change after rebalancing (the output of clusters).